### PR TITLE
Fix Air Lock and Cloud Nine

### PR DESCRIPTION
**Pull request by WreckDra**
Moved them from AnyModifyPokemon to AnyUpdate base on information in the [Air Lock and Cloud Nine Difficulties](https://github.com/Zarel/Pokemon-Showdown/issues/1789) issue from @gamut-was-taken

### CODE REVIEW

**Comment by ascriptmaster** [comment]
Have you confirmed that this works as expected?

**Comment by ascriptmaster** [comment]
The function that calls `ModifyPokemon` runs once per target when using a multi-target move in Doubles/Triples, and that function resets `ignore` flags when it is run. This means that damage calcs might be wrong if you use this code, if you use Lava Plume in Sun while a Cloud Nine Pokemon is on the field.

At least, that's what I remember.

**Comment by WreckDra** [comment]
It works with residual damage from weather as intended. You're right about the damage calcs though. We could just .ignore twice I guess. Once  onAnyUpdate, and once onModifyPokemon if no one can think of a better way to condense it. We need it to be run before TryWeather at least once, though.

**Comment by ascriptmaster** [comment]
Nah, to save cycles you might be able to put the other one on SwitchIn because that's where the glitch is. When a Cloud Nine Mon switches in, set all active Pokemon's ignore['WeatherTarget']

**Comment by ascriptmaster** [comment]
I was actually thinking of a solution earlier that removes the need for Cloud Nine to even use the ignore keyword in the first place, but I haven't gotten my version working yet.

**Comment by WreckDra** [comment]
We could just have Weather return null in place of the ignore if we really wanted. That I feel like that would somehow mess with upkeep etc. even though those are handled before Weather (PS is sometimes really unpredictable with this sort of thing). I am curious as to your build though, I have no way to test my builds anymore though due to a BAM's server owner being more or less dead for a little while at least.

**Comment by ascriptmaster** [comment]
My idea was to set an ignore flag on the weather directly and deal with things that way. Cloud Nine and Air Lock can then set or unset it on switch in and switch out / ability removal.

**Comment by ascriptmaster** [comment]
Oh wait that won't work if a weather is set after the Cloud Nine Mon is active

**Comment by WreckDra** [comment]
I might try making Weather return null actually, since all it handles is sand / hail damage with Solar Power and Dry Skin / Rain Dish etc damage / healing. I have no way of testing this as of right now though, and I am uncomfortable adding it to the request until I know how it effects the other elements of the weather (upkeep, etc.). If you have a place to test making Weather return null, knock yourself out. :)
I would like to push this through before monday so it doesnt end up affecting UPL at all (Rayquaza and weather are used a ton here).

**Comment by Zarel** [comment]
`ModifyPokemon` should not exist, but neither is `Update` an appropriate time. This really just needs to be refactored into something that isn't run all the time.

**Comment by WreckDra** [comment]
It looks like ```ModifyPokemon``` is used for a few abilities to not affected be by Mold Breaker and the likes as well as other things. If you do wish to remove it though, we could always do something like this:
```javascript
onWeather: {
 if (pokemon.ability === 'Air Lock')
 return null;
 },
```
I don't know if we need to check for the ability, but i figured it would be better safe than sorry. I think null is the better return though due to canceling weather all together potentially screwing up other things. Of course this would be duplicated for Cloud Nine. Another option is to do this sort of thing to all of the weathers in status.js, but that would be a bunch of unnecessary lines I feel.

**Comment by Zarel** [comment]
Yes, I'm aware; we're slowly working through refactoring everything not to use `ModifyPokemon`. In the meantime, we should stop using it.

> `if (pokemon.ability === 'Air Lock')`

In theory, we have the `TryWeather` event for that.

**Comment by ascriptmaster** [comment]
Oh, that's a great idea, simply have weather effects invoke `TryWeather`. :+1:

**Comment by WreckDra** [comment]
Well um... LOL. I went digging in the Battle Engine and found ```TryWeather``` and frankly this bug is making less and less sense because of [this](https://github.com/Zarel/Pokemon-Showdown/blob/e22d0a5cebd330e8e79022e860701f8f951c6a74/battle-engine.js#L1802-L1803). According to logic, ```Weather``` should return ```false``` before it has a chance to return any other Weather Effect in Status.js. I figured just adding ```pokemon.ignore['WeatherTarget'] = true;``` to ```SwitchIn``` would probably be our best bet of catching all jagged edges regarding both events unless ```TryWeather``` is bugged altogether.

**Comment by ascriptmaster** [comment]
Why is that confusing? `effectiveWeather` only asks for the current weather (or blank if Cloud Nine is active) but weather-based events will still run if they are not suppressed [within the runEvent code](https://github.com/Zarel/Pokemon-Showdown/blob/master/battle-engine.js#L2046-2049).

Remember, the `ignore` flags are scrubbed at every cycle of `ModifyPokemon`, which is... problematic. A volatile could work too, if we hardcode the events system to check for it. Just make sure you get all the use-cases for when it should be added/removed from a Pokemon.

**Comment by Zarel** [comment]
We need to get rid of the `ignore` events (replace with `Try*`) anyway...

**Comment by ascriptmaster** [comment]
Running events within events sounds dangerous...

**Comment by Zarel** [comment]
We already do that all the time.

**Comment by Zarel** [comment]
Don't worry, we have safeguards to prevent stack overflows.

**Comment by ascriptmaster** [comment]
So ignoring items would be a `TryItem` event? That would work.

For ability ignores we might be able to hardcode for Gastro Acid and Mold Breaker, or have events for those anyhow.

I think that's all the `ignore`s that get used.

**Comment by WreckDra** [comment]
Would creating a new ```ignoreWeather``` event be the way to go then?

**Comment by WreckDra** [comment]
lol, misclicked close

**Comment by Zarel** [comment]
> Running events within events sounds dangerous...

Actually, one thing we _should_ be very careful about is never to run a `Try` event from inside another one. If it ever comes up, we might want to hardcode the interaction.

**Comment by Zarel** [comment]
> Would creating a new ```ignoreWeather``` event be the way to go then?

In theory, the current `TryWeather` event should be sufficient. It's still unclear to me what the problem with it is.

**Comment by ascriptmaster** [comment]
@WreckDra, if you're going to be having Cloud Nine only have the `TryWeather`, you need to fix this line https://github.com/Zarel/Pokemon-Showdown/blob/master/battle-engine.js#L2046 so that instead of checking the ignore, it runs the event. Something like this:

    if (effect.effectType === 'Weather' && !this.runEvent('TryWeather', target))

Then try that and see if everything works.

**Comment by WreckDra** [comment]
Alright, I will try and figure out what ```BattleStatuses``` are actually defined as after I sleep so my fried brain doesn't try anything silly.

**Comment by Zarel** [comment]
You probably want to do `npm test` before pushing. It'll tell you the errors that Travis CI is currently complaining about.

File: battle-engine.js [hunk 1]
```diff
@@ -2043,7 +2043,7 @@ Battle = (function () {
 			this.debug(eventid + ' handler suppressed by Gastro Acid, Klutz or Magic Room');
 			return relayVar;
 		}
-		if (target.ignore && target.ignore[effect.effectType + 'Target']) {
+		if (effectType === 'Weather' && eventid !== 'TryWeather' && !this.runEvent('TryWeather', target)) {
```
Comment (ascriptmaster) [inline battle-engine.js]: `line 2046  col 13  'effectType' is not defined.`

It should be `effect.effectType`, I think. Also, make sure you test the code yourself before you push.

**Comment by ascriptmaster** [comment]
Are you still working on this?

**Comment by ascriptmaster** [comment]
Redundant because of #1855

**Pull request closed without merging**